torch/optim/lr_scheduler.py

```diff
@@ -777,8 +971,14 @@ def step(self, epoch=None):
                 self.T_i = self.T_0
                 self.T_cur = epoch
         self.last_epoch = math.floor(epoch)
+
+        self._get_lr_called_within_step = True
         for param_group, lr in zip(self.optimizer.param_groups, self.get_lr()):
             param_group['lr'] = lr
+        self._get_lr_called_within_step = False
```

Comment:
Wouldn't a more normal way to implement this to just have `get_latest_lr` throw no warning (and `get_lr` point at it with a warning), so you don't have to muck around with this attribute?

Reply:
If I understand correctly, you suggesting to just make `get_lr` throw an error when called? `get_lr` is always called within `step`, and we don't want an error thrown in this case. This attribute is meant to check that `get_lr` is called within `step` and avoid throwing an error in this case. Is that what you meant?

Reply:
> If I understand correctly, you are suggesting to just make get_lr throw an error when called?

If by "throw an error" you mean "raise a warning", yes, I am suggesting that. However, I am *also* suggesting that you change the line: 

```
        for param_group, lr in zip(self.optimizer.param_groups, self.get_lr()):
```

To be:

```
        for param_group, lr in zip(self.optimizer.param_groups, self.get_lr_nowarn()):
```

where `get_lr_nowarn` is a version of `get_lr` that has the same semantics, but doesn't do a warning.

Reply:
> However, I am _also_ suggesting that you change the line:
> 
> ```
>         for param_group, lr in zip(self.optimizer.param_groups, self.get_lr()):
> ```
> 
> To be:
> 
> ```
>         for param_group, lr in zip(self.optimizer.param_groups, self.get_lr_nowarn()):
> ```
> 
> where `get_lr_nowarn` is a version of `get_lr` that has the same semantics, but doesn't do a warning.

I see, so you are proposing (1) `step` uses `get_lr_no_warning` which computes learning rate without warning, (2) `get_lr` raises a warning and calls `get_lr_no_warning`.

However, users in the wild already overwrite `get_lr` with their own, and expect `step` to invoke it without modification. What I describe would then break those codes. Thoughts?

> I think another way of doing this would be with a context manager?

Fair :)

Reply:
> Fair :) I'll use that.

Context manager is a strict improvement because it will handle exceptions correctly.

Reply:
> However, users already overwrite get_lr with their own, and expect step to invoke it without modification. What I described just now would then break those codes. Thoughts?

Err, if this is true, you're not going to get warnings in this case, because they've overridden your warnings? It is getting a bit hard to keep track of all the cases. Is there a clear description of how these changes interact with user defined schedulers?

Reply:
> Err, if this is true, you're not going to get warnings in this case, because they've overridden your warnings? It is getting a bit hard to keep track of all the cases. Is there a clear description of how these changes interact with user defined schedulers?

I don't need to provide a warning in this case. My goal is just to alert to the changes that happened.

* Users should be alerted that some existing schedulers changed from closed form to chainable form.
* Users should be encouraged to use `get_last_lr` to get the last computed value instead of `get_lr` (e.g. by the warning discussed here), since the modified `get_lr` wouldn't returned the last lr for chainable schedulers. I can't make assumptions on how this would behave for a user-modified scheduler.
* Users should be discouraged from providing epoch when not needed.
* Users modified schedulers should continue to work as expected.

> Context manager is a strict improvement because it will handle exceptions correctly.

Agree.